File: src/pwArg.ts
import { decodeEnvArgs } from './env.js';
import { parseBooleanValue } from './helper.js';
import type { CustomArgs, CustomArgValue } from './types.js';

export interface ReadOptions<T> {
  /** Value returned when the argument was not passed. Without it, a missing argument throws. */
  readonly default?: T;
}

export interface PwArg {
  /** Returns a defensive copy of every custom argument. */
  raw(): CustomArgs;
  /** Returns whether the argument was passed. */
  has(name: string): boolean;
  /** Reads the argument as a string, using the last value when repeated. */
  string(name: string, options?: ReadOptions<string>): string;
  /** Reads the argument as a finite number, using the last value when repeated. */
  number(name: string, options?: ReadOptions<number>): number;
  /** Reads a bare flag or a `true`/`false` value, using the last value when repeated. */
  boolean(name: string, options?: ReadOptions<boolean>): boolean;
  /** Reads every value of a repeated argument, or a single-item array for one value. */
  array(name: string, options?: ReadOptions<readonly string[]>): string[];
}

type ScalarValue = string | boolean;

export function createPwArg(args: CustomArgs = decodeEnvArgs()): PwArg {
  const lookup = (name: string): CustomArgValue | undefined => (Object.hasOwn(args, name) ? args[name] : undefined);

  return {
    raw: () => cloneArgs(args),
    has: (name) => lookup(name) !== undefined,
    string: (name, options = {}) => readString(name, lookup(name), options.default),
    number: (name, options = {}) => readNumber(name, lookup(name), options.default),
    boolean: (name, options = {}) => readBoolean(name, lookup(name), options.default),
    array: (name, options = {}) => readArray(lookup(name), options.default),
  };
}

export const pwArg: PwArg = createPwArg();

function readString(name: string, value: CustomArgValue | undefined, fallback: string | undefined): string {
  if (value === undefined) return requireFallback(name, fallback);

  return String(lastValue(name, value));
}

function readNumber(name: string, value: CustomArgValue | undefined, fallback: number | undefined): number {
  if (value === undefined) return requireFallback(name, fallback);

  const scalar = lastValue(name, value);
  const parsed = typeof scalar === 'string' ? Number(scalar) : Number.NaN;
  if (!Number.isFinite(parsed)) throw new Error(`Custom argument "${name}" must be a number`);

  return parsed;
}

function readBoolean(name: string, value: CustomArgValue | undefined, fallback: boolean | undefined): boolean {
  if (value === undefined) return requireFallback(name, fallback);

  const parsed = parseBooleanValue(lastValue(name, value));
  if (parsed === undefined) throw new Error(`Custom argument "${name}" must be a boolean`);

  return parsed;
}

function readArray(value: CustomArgValue | undefined, fallback: readonly string[] | undefined): string[] {
  if (value === undefined) return [...(fallback ?? [])];
  if (Array.isArray(value)) return [...value];
  return [String(value)];
}

function lastValue(name: string, value: CustomArgValue): ScalarValue {
  if (!Array.isArray(value)) return value;

  const last = value.at(-1);
  if (last === undefined) throw requiredError(name);

  return last;
}

function requireFallback<T>(name: string, fallback: T | undefined): T {
  if (fallback === undefined) throw requiredError(name);

  return fallback;
}

function requiredError(name: string): Error {
  return new Error(`Custom argument "${name}" is required`);
}

function cloneArgs(args: CustomArgs): CustomArgs {
  return Object.fromEntries(
    Object.entries(args).map(([name, value]) => [name, Array.isArray(value) ? [...value] : value]),
  );
}

File: src/pw_arg.rs
//! Port of `pwArg.ts`.

use crate::env::decode_env_args;
use crate::helper::parse_boolean_value;
use crate::parse_cli::{ArgValue, CustomArgs};
use std::sync::LazyLock;

/// Typed reader over custom arguments (the `createPwArg` return value).
pub struct PwArg {
    args: CustomArgs,
}

/// Process-wide singleton, lazily initialized once from `PLAYWRIGHT_ARGS_JSON`
/// (the `export const pwArg = createPwArg()` equivalent).
static PW_ARG: LazyLock<PwArg> =
    LazyLock::new(|| PwArg::from_env().unwrap_or_else(|error| panic!("{error}")));

pub fn pw_arg() -> &'static PwArg {
    &PW_ARG
}

impl PwArg {
    pub fn new(args: CustomArgs) -> Self {
        Self { args }
    }

    pub fn from_env() -> Result<Self, String> {
        Ok(Self::new(decode_env_args()?))
    }

    pub fn raw(&self) -> CustomArgs {
        self.args.clone()
    }

    pub fn has(&self, name: &str) -> bool {
        self.args.contains_key(name)
    }

    pub fn string(&self, name: &str, default: Option<&str>) -> Result<String, String> {
        match self.scalar_value(name)? {
            Some(value) => Ok(value.into_string()),
            None => default.map(str::to_string).ok_or_else(|| required(name)),
        }
    }

    pub fn number(&self, name: &str, default: Option<f64>) -> Result<f64, String> {
        let Some(value) = self.scalar_value(name)? else {
            return default.ok_or_else(|| required(name));
        };
        let ArgValue::Str(value) = value else {
            return Err(number_error(name));
        };

        let parsed: f64 = value.trim().parse().map_err(|_| number_error(name))?;
        if !parsed.is_finite() {
            return Err(number_error(name));
        }
        Ok(parsed)
    }

    pub fn boolean(&self, name: &str, default: Option<bool>) -> Result<bool, String> {
        let value = match self.scalar_value(name)? {
            Some(value) => value,
            None => return default.ok_or_else(|| required(name)),
        };

        parse_boolean_value(&value).ok_or_else(|| format!("Custom argument \"{name}\" must be a boolean"))
    }

    pub fn array(&self, name: &str, default: Option<&[String]>) -> Vec<String> {
        match self.args.get(name) {
            None => default.unwrap_or_default().to_vec(),
            Some(ArgValue::List(values)) => values.clone(),
            Some(value) => vec![value.clone().into_string()],
        }
    }

    /// The last repeated value wins for scalar readers; `Ok(None)` means missing.
    fn scalar_value(&self, name: &str) -> Result<Option<ArgValue>, String> {
        match self.args.get(name) {
            None => Ok(None),
            Some(ArgValue::List(values)) => match values.last() {
                Some(last) => Ok(Some(ArgValue::Str(last.clone()))),
                None => Err(required(name)),
            },
            Some(value) => Ok(Some(value.clone())),
        }
    }
}

fn required(name: &str) -> String {
    format!("Custom argument \"{name}\" is required")
}

fn number_error(name: &str) -> String {
    format!("Custom argument \"{name}\" must be a number")
}
